File: leetcode/ValidParentheses.java
/**
 * Valid Parentheses
 * https://leetcode.com/problems/valid-parentheses/
 */
class Solution {
    public boolean isValid(String s) {
       Stack<Character> pancakes = new Stack<Character>();
        
        for (Character x : s.toCharArray()) {
            switch (x) {
                case '(' -> pancakes.push(')');
                case '[' -> pancakes.push(']');
                case '{' -> pancakes.push('}');
                default -> {
                     if (pancakes.isEmpty() || pancakes.pop() != x) return false;
                }
            }
        }
        
        return pancakes.isEmpty();
    }
}

File: leetcode/ValidParentheses.js
/**
 * https://leetcode.com/problems/valid-parentheses/
 * 
 * @param {string} s
 * @return {boolean}
 */
var isValid = function(s) {
    let stack = [];
    let characters = s.split('');

    for (character of characters) {
        switch (character) {
            case '(':
                stack.push(')');
                break;
            case '[':
                stack.push(']');
                break;
            case '{':
                stack.push('}');
                break;
            default:
                if (stack.length === 0 || stack.pop() !== character) {
                    return false;
                }
        }
    };

    return stack.length === 0;
};

console.log(isValid('()')); // true
console.log(isValid('()[]{}')); // true
console.log(isValid('(]')); // false
